### ISSUE TITLE
improve `from_audio` loading speed

### ISSUE THREAD

**Issue by faroit** [tensorflow/io#647]
I am currently [evaluating](https://github.com/faroit/python_audio_loading_benchmark/pull/8/files#diff-76e989ed1c4ff81f51d6a966464d71af) the performance on tfio for fast wav loading compared to other libs. 

I am using the approach as proposed in #615 by @yongtang. Compared to python function loaders such as scipy.io or soundfile, the native tensorflow ops are significantly slower (even though they do not need to convert from/to numpy first). 

![benchmark_tf](https://user-images.githubusercontent.com/72940/69014652-c7c6fe00-098c-11ea-95a0-8db114cde37f.png)

It would be great if the performance of the tfio audio ops could further be improved

**Comment by yongtang** [comment]
Thanks @faroit. I am currently running the test myself. Will be back soon.

One thing I want to note, is that in PR #613 that was merged a weeks ago, we enabled AVX and SSE4.2, which might have a positive impact on performance.

I also notice in your branch, the `tensorflow-io==0.10.0` was specified in `requirement.txt`. Since  tensorflow-io==0.10.0 has not been released, Not sure which version you are using.

I will just use the latest tensorflow-io-nightly to take a look.

**Comment by yongtang** [comment]
Thanks @faroit for the work. I tried to run the branch `tf2_librosa7`, but seems to get a different result.

I did make several changes in the code, as `ar_mad` does not work inside the container. Also I only enabled `wav`.

Here are the places I made the change:

In benchmark_tf.py, comment out ar_mad, also use `loaders.load_tfio_fromaudio` if lib is `tfio_fromaudio`:
```diff
diff --git a/benchmark_tf.py b/benchmark_tf.py
index 6fd907a..e8d3bfa 100644
--- a/benchmark_tf.py
+++ b/benchmark_tf.py
@@ -50,7 +50,7 @@ if __name__ == "__main__":
     libs = [
         'ar_gstreamer',
         'ar_ffmpeg',
-        'ar_mad',
+        #'ar_mad',
         'aubio',
         'pydub',
         'soundfile',
@@ -72,11 +72,16 @@ if __name__ == "__main__":
                 )
 
                 dataset = tf.data.Dataset.from_tensor_slices(audio_files)
-                if lib in ["tf_decode_wav", "tfio_fromaudio"]:
+                if lib in ["tf_decode_wav"]:
                     dataset = dataset.map(
                         lambda x: loaders.load_tf_decode_wav(x, args.ext),
                         num_parallel_calls=1
                     )
+                elif lib in ["tfio_fromaudio"]:
+                    dataset = dataset.map(
+                        lambda x: loaders.load_tfio_fromaudio(x),
+                        num_parallel_calls=1
+                    )
                 else:
                     loader_function = getattr(loaders, 'load_' + lib)
                     dataset = dataset.map(
```

In plot.py, only test wav:
```diff
diff --git a/plot.py b/plot.py
index 02e1418..e4e485c 100644
--- a/plot.py
+++ b/plot.py
@@ -3,9 +3,9 @@ import seaborn as sns
 import matplotlib.pyplot as plt
 
 
-for package in ['pytorch', 'tf', 'np', 'metadata']:
+for package in ['tf']:
     dfs = []
-    for ext in ["wav", "mp3", "mp4", "ogg", "flac"]:
+    for ext in ["wav"]:
         dfs.append(pd.read_pickle("results/benchmark_%s_%s.pickle" % (package, ext)))
 
     df = pd.concat(dfs, ignore_index=True)
```

In run.sh, only invoke wav:
```diff
diff --git a/run.sh b/run.sh
index 35e01f5..e2d539e 100644
--- a/run.sh
+++ b/run.sh
@@ -1,11 +1,11 @@
 ## declare an array variable
-declare -a exts=("wav" "mp3" "mp4" "ogg" "flac")
+declare -a exts=("wav")
 
 ## now loop through the above array
 for i in "${exts[@]}"
 do
-    python benchmark_np.py --ext "$i"
-    python benchmark_pytorch.py --ext "$i"
+    #python benchmark_np.py --ext "$i"
+    #python benchmark_pytorch.py --ext "$i"
     python benchmark_tf.py --ext "$i"
-    python benchmark_metadata.py --ext "$i"
+    #python benchmark_metadata.py --ext "$i"
 done
```

In loader.py, uncomment audioread:
```diff
diff --git a/loaders.py b/loaders.py
index bcf9d19..22c3591 100644
--- a/loaders.py
+++ b/loaders.py
@@ -1,6 +1,6 @@
 from scipy.io import wavfile
 import audioread.rawread
-# import audioread.gstdec
+import audioread.gstdec
 import audioread.maddec
 import audioread.ffdec
 import matplotlib.pyplot as plt
```

In requirements.txt, change to tensorflow-io-nightly (as tensorflow-io 0.10.0 has not been released yet):
```diff
diff --git a/requirements.txt b/requirements.txt
index f89852b..99874aa 100644
--- a/requirements.txt
+++ b/requirements.txt
@@ -8,7 +8,7 @@ pydub==0.23.1
 sox==1.3.7
 torchaudio==0.3.1
 aubio==0.4.9
-tensorflow_io==0.10.0
+tensorflow-io-nightly==0.10.0.dev2227
 PyGObject
 seaborn
 pandas

```

Here is the png file I got:

![benchmark_tf](https://user-images.githubusercontent.com/6932348/69019291-d699ba80-0964-11ea-92d6-36242581e0c3.png)

**Comment by yongtang** [comment]
This seems to be quite different, let me try again.

**Comment by yongtang** [comment]
This is another run. I removed aubio, ar_gstreamer, ar_ffmpeg as they takes longer and may skew the ratio.

There are some variations with respect to librosa, scipy, and tf_decode_wav, but not by much:

![benchmark_tf](https://user-images.githubusercontent.com/6932348/69019493-b8808a00-0965-11ea-8300-038519561e19.png)

**Comment by yongtang** [comment]
In the mean time, here is the test setup:

My test is run on Ubuntu 18.04 AMI, with 4CPU+16GB (CPU is AMD).

I run the test inside a docker container.

`Docker version: 19.03.4`

During the `docker run`, I use `--net=host` to get directory to host network stack.

**Comment by yongtang** [comment]
Here are some of the possible factors I can think of for the difference:
1) The benchmark_tf.py was changed, wondering during your test if this was enabled?
2) As I mentioned, we enabled SSE 4.2 and AVX a week ago (in PR #613). If your local install was not updated, then it might cause quite some difference.
3) I don't think docker usage or version could be an issue, though are you using default bridge network mode or host mode?

**Comment by faroit** [comment]
@yongtang thanks for testing this. I wasn't expecting that you actually run my tests. I would have included more info for you how to run it. So thanks again for all your effort. I can only repeat how valuable and important a fast and seekable audio loader is for audio (especially music and environmental sounds), therefore 👍 for all your effort so far.

> As I mentioned, we enabled SSE 4.2 and AVX a week ago (in PR #613). If your local install was not updated, then it might cause quite some difference.

I used the latest 0.10.0 nightly, so #613 was already in there, I guess. Maybe the difference stems from that I ran it on OS X. I will try again on a different node

> I don't think docker usage or version could be an issue, though are you using default bridge network mode or host mode?

I didn't ran it inside docker yet, will try this

> +tensorflow-io-nightly==0.10.0.dev2227

alright. Since this is just a WIP branch, i will just use `tensorflow-io-nightly` until `0.10.0` is released.

Two more things 

1. is `from_ffmpeg` also ready to read audio in graph mode at this point?
2. as you have noticed, the benchmark script iterates over multiple formats. What is the recommended way to ignore unsupported files in that pipeline? `tf.data.experimental.ignore_errors()` ?

**Comment by yongtang** [comment]
@faroit Ah I realized that SSE 4.2 and AVX may not be enabled on macOS yet. Let me double check.

**Comment by yongtang** [comment]
@faroit There are quite a few optimizations that might be done. #649 is opened (not all covered yet). As I am in kubecon this week so things might be slowed. I will try to update PR if I can find some time.

**Comment by yongtang** [comment]
@faroit Ogg supported is tracked in #650

**Comment by yongtang** [comment]
@faroit With the release of tensorflow-io 0.10.0, SSE 4.2 and AVX are enabled on both Linux and macOS platforms. This should significantly improve the speed loading audio files. Would you mind rerun the test and check again?

**Comment by faroit** [comment]
![benchmark_tf](https://user-images.githubusercontent.com/72940/70336554-9fbf1200-1849-11ea-8b2f-a99b9ac829dc.png)

looks way better now. great work! thanks

**Comment by terrytangyuan** [comment]
Perhaps we should add regression test for this to make sure the speed does not decrease? @yongtang

**Comment by yongtang** [comment]
Thanks @faroit for the test! Let me know if you encounter any other issues.

@terrytangyuan Yes we should setup a benchmark test. I think adding pytest-benchmark will help. We also want to standardize the audio input (only support WAV but OGG is being added in #662, and I am looking into other audio types as well) across the board.

**Comment by faroit** [comment]
> @terrytangyuan Yes we should setup a benchmark test. I think adding pytest-benchmark will help. We also want to standardize the audio input (only support WAV but OGG is being added in #662, and I am looking into other audio types as well) across the board.

@yongtang @terrytangyuan I would propose to look into FLAC next.
